repo readme update (#1049)
brimsec/zq: cmd/zq/README.md

# `zq`

`zq` is a command-line tool to search, analyze, and transform structured logs. 
 It evaluates [ZQL ](../../zql/docs/README.md) queries against input log

>   files, producing an output log stream in the [ZNG](../../zng/docs/spec.md)
>   format by default. For all `zq` options, see the built-in help by running:

Reviewer: ```suggestion
 files, producing an output log stream in the [ZNG](../../zng/docs/spec.md)
 format by default. For all `zq` options, see the built-in help by running
```

```
zq help
```

## Examples

> Here are a few examples using a small Zeek formatted log file, `conn.log
> `, located in this directory. See the

Reviewer: ```suggestion
Here are a few examples using a small Zeek formatted log file, `conn.log`,
located in this directory. See the
```

> Here are a few examples using a small Zeek formatted log file, `conn.log
> `, located in this directory. See the

Reviewer 2: ```suggestion
Here are a few examples using a small Zeek formatted log file, `conn.log`,
located in this directory. See the
```

When the backtick was on its own line, it formatted this as `conn.log `

[zq-sample-data repo](https://github.com/brimsec/zq-sample-data) for more test
data, which is used in the examples in the
[query language documentation](../../zql/docs/README.md).

To cut the columns of a Zeek "conn" log like `zeek-cut`, and output to the

>  terminal run:

Reviewer: ```suggestion
 terminal run
```

```
zq -t "* | cut ts,id.orig_h,id.orig_p" conn.log
```

The `-t` tells `zq` to use [TZNG](../../zng/docs/spec.md#4-zng-text-format-tzng)
for its output format. The "`*`" 
tells `zq` to match every line, which is sent to the `cut` processor
using the UNIX-like pipe syntax.

When looking over everything like this, you can omit the search pattern

> as a shorthand and simply type:

Reviewer: ```suggestion
as a shorthand and simply type
```

```
zq -t "cut ts,id.orig_h,id.orig_p" conn.log
```

> The default output is the binary ZNG format.  If you want just the tab-separated

Reviewer: ```suggestion
The default output is the binary ZNG format. If you want just the tab-separated
```

>  lines like `zeek-cut`, you can specify text output:

Reviewer: ```suggestion
lines like `zeek-cut`, you can specify text output:
```

```
zq -f text "cut ts,id.orig_h,id.orig_p" conn.log
```
If you want the old-style Zeek [ASCII TSV](https://docs.zeek.org/en/stable/examples/logs/)
log format, run the command with the `-f` flag specifying `zeek` for the output

> format:

Reviewer: ```suggestion
format.
```

```
zq -f zeek "cut ts,id.orig_h,id.orig_p" conn.log
```
You can use an aggregate function to summarize data over one or
more fields, e.g., summing field values, counting, or computing an average.
```
zq -t "sum(orig_bytes)" conn.log
zq -t "orig_bytes > 10000 | count()" conn.log
zq -t "avg(orig_bytes)" conn.log
```

> The [ZNG specification](../../zng/docs/spec.md) describes the significance of
>  the
> `_path` field.  By leveraging this, diverse Zeek logs can be combined into a single
> file.

Reviewer: ```suggestion
The [ZNG specification](../../zng/docs/spec.md) describes the significance of
the `_path` field. By leveraging this, diverse Zeek logs can be combined into
a single file.
```

> The [ZNG specification](../../zng/docs/spec.md) describes the significance of
>  the
> `_path` field.  By leveraging this, diverse Zeek logs can be combined into a single
> file.

Reviewer 2: ```suggestion
The [ZNG specification](../../zng/docs/spec.md) describes how the format can
represent a stream of heterogeneously typed records. By leveraging this,
diverse Zeek logs can be combined into a single file.
```

This was content that you didn't change, but during this review pass I realize it's gone stale. I think this suggestion brings it current.

```
zq *.log > all.zng
```

### Comparisons

> Revisiting the `cut` example shown above:

Reviewer: ```suggestion
Consider again the `cut` example shown above, which is
```

```
zq -f text "cut ts,id.orig_h,id.orig_p" conn.log
```

> This is functionally equivalent to the `zeek-cut` command-line:

Reviewer: ```suggestion
This is functionally equivalent to following `zeek-cut` command.
```

```
zeek-cut ts id.orig_h id.orig_p < conn.log
```

If your Zeek events are stored as JSON and you are accustomed to querying with `jq`,

> the equivalent would be:

Reviewer: ```suggestion
the equivalent would be
```

```
jq -c '. | { ts, "id.orig_h", "id.orig_p" }' conn.ndjson
```

Comparisons of other simple operations and their relative performance are described
at the [performance](../../performance/README.md) page.

## Formats

| Format | Read | Auto-Detect | Write | Description |
|--------|------|-------------|-------|-------------|
| zng | yes | yes | yes | [ZNG specification](../../zng/docs/spec.md) |
| tzng | yes | yes | yes | [TZNG specification](../../zng/docs/spec.md#4-zng-text-format-tzng) |
| ndjson | yes | yes | yes | Newline delimited JSON records |
| zeek  | yes | yes | yes | [Zeek compatible](https://docs.zeek.org/en/stable/examples/logs/) tab separated values |

> | zjson | yes | yes | yes | Zeek JSON |

Reviewer 2: While many readers may not linger on this, strictly speaking, I think it would be inaccurate to call this "Zeek JSON", since users coming from the Zeek world are more likely to see this as a reference to [Zeek's regular ASCII logger outputting JSON](https://docs.zeek.org/en/current/frameworks/logging.html#ascii-writer) or the use of the [json-streaming-logs](https://github.com/corelight/json-streaming-logs) package. I don't really feel like I'm well-suited to say what we should say instead since my understanding is that this zjson format is something that was effectively created for internal developer use, and I'm not a developer. My suspicion is that it would more accurately be called "ZNG over JSON" and link to https://github.com/brimsec/zq/blob/zq-readme-update/zng/docs/zng-over-json.md (which I believe is a woefully out-of-date doc, alas).

```

is functionally equivalent to this `zeek-cut` command:

```
zeek-cut ts id.orig_h id.orig_p < conn.log
```

If your Zeek events are stored as JSON, the equivalent `jq` command is:

```
jq -c '. | { ts, "id.orig_h", "id.orig_p" }' conn.ndjson
```

Comparisons of other simple operations and their relative performance are described
at the [performance](../../performance/README.md) page.

## Formats

| Format | Read | Auto-Detect | Write | Description |
|--------|------|-------------|-------|-------------|
| zng | yes | yes | yes | [ZNG specification](../../zng/docs/spec.md) |
| tzng | yes | yes | yes | [TZNG specification](../../zng/docs/spec.md#4-zng-text-format-tzng) |
| ndjson | yes | yes | yes | Newline delimited JSON records |
| zeek  | yes | yes | yes | [Zeek compatible](https://docs.zeek.org/en/stable/examples/logs/) tab separated values |
| zjson | yes | yes | yes | [ZNG over JSON](../../zng/docs/zng-over-json.md) |
| parquet | yes | no | no | [Parquet file format](https://github.com/apache/parquet-format#file-format)
| table | no | no | yes | table output, with column headers |
| text | no | no | yes | space separated output |

> | types | no | no | * | Used to output text description of read record types |

Reviewer 2: Nit: My eyes didn't notice this one in my first review pass. The "*" is not super clear to me in what it's communicating. In a table full of yes/no values, I'd be inclined to put "N/A" or something of the sort.

Author: ah, thanks! I meant to handle the one specially, will address.
